### DISCUSSION TITLE
Making `Color = Color.` cases consistent and well-established

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#7222] [Language Ideas]
In spite of:
- encountering the compiler error of [this issue](https://github.com/dotnet/roslyn/issues/45739)
- and fixing it via [this PR](https://github.com/dotnet/roslyn/pull/68283)

I discovered there are some inconsistencies with how the compiler tries to resolve `Color = Color.` cases, as they are commonly referred to. Without knowing exactly the spec and how it handles those cases, those inconsistencies are there and for no good reason.

First of all, a simple [repro](https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoBAGABAYQFcBnAFwHsx8ALCAJ1wF5cEUA6G+gbkxwJIpgAchDABTYgAcIAYzE8MvAMy4xMQmALkANuQYBvXACUxAEwA0uAOJ0xa3AF9MylkgKZ9mXN5YqZ5GDItXQZ8HT0mYL02E1MFH19cf0DSFnYASRhSJTdWSNY2TOykNgBZCAAPADUIbUJ5Lx8EPwCggCFyHTEIGFwOrp7I/u1umDYAFTp6gGVSOlgAc3imlpT+MkpOUIFN2gZmIg2qPbLKmrqGjATG72a0gDZcWwhTAO0AT1xYVMPBEXEpLIxJFfpR/hJpHI2NNKGJ8NoIMRiGV3qCwOd6lxcJgnIosCpkLgACIeG6JViPZ6vGAfNLQuaLNL5dizeYwBZsACiYEkpHey1uBJQlO61NpaPBgKhMPE8MRxHWf1EEKBIJ2wmVUrE0NhcqRhRSPTkAvJwqeorenxlcIRSNw1r1CuYDttyMyZCNl2uVxW5pelqioXC+0DMTMClxmBgmshwIlMaBpJ9gtc9t1rrJnmTCVNIv9NKt6fluHdpE9kRgYgA7gAKACUJu9Od9ySC31wpVR6oxwOYSgALI2fGSEncXcWhGIKqlGAA+XDqbTaIfeEe+hD9jtiUjUcimGvt8d251F/WdtE9utr7xZ5veXEJXGR/GqdSaACCSYSbSGbAA8pW5g4s4L5qBofRfj4AG9rgKBARguJAA==).

The above code shows a lot of different cases, while still probably missing a few, that clearly showcase a fundamental confusion with what is allowed to be declared with a `Color = Color.` pattern and what is not.

## The Rules

Currently, the rules of resolving `Color = Color.` as a type qualification are as follows:
- fine on method locals
- fine on method constants
- fine on method parameters
- fine on non-constant type fields
- **not fine** on constant type fields
- **not fine** on enum fields (semantically the same as the above)

All the above cases also include the possibility of type aliases.

However, there is partial recognition for namespaces, with the rules of resolving `Color = Color.` as a namespace qualification being:
- **not fine** on method locals
- **not fine** on method constants
- fine on method parameters
- **not fine** on non-constant type fields
- **not fine** on constant type fields
- **not fine** on enum fields (semantically the same as the above)

## Analysis

A massive portion of the above cases are completely pathogenic. However, the problem was discovered via the restriction of the pattern when assigning constant fields in a type, which is seemingly a bug in resolving the case.

If that is to be fixed, the spec then needs to state exactly how to treat those patterns and how to resolve the identifier.

## Proposal

In all of the above cases, we are talking about initialization, **not** reassignment. In none of the above cases do we want to directly access the symbol under initialization, as it would result in either an unassigned symbol, or a default value being passed. Nobody sane would ever care about the default value of an unassigned value upon its assignment.

Pathogenic as the other similar cases may be, a consistency would be very desirable, to completely disregard the symbol under initialization.

And finally, we will be able to type
```csharp
public const AttributeTargets AttributeTargets = AttributeTargets.Class | AttributeTargets.Interface;
```

**Comment by CyrusNajmabadi** [comment]
> However, there is partial recognition for namespaces

There should be no recognition for this.  The rules about `Color Color` [state specifically](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#11762-identical-simple-names-and-type-names)

> In a member access of the form E.I, if E is a single identifier, and if the meaning of E as a simple_name ([§11.7.4](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#1174-simple-names)) is a constant, field, property, local variable, or parameter with the same type as the meaning of E as a type_name ([§7.8.1](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/basic-concepts.md#781-general)), then both possible meanings of E are permitted.

`namespace` is not a "constant, field, property, local variable, or parameter".  And namespaces do not have types at all, and thus cannot satisfy the "same type as the meaning" clause of the above.

**Comment by CyrusNajmabadi** [comment]
> I discovered there are some inconsistencies with how the compiler tries to resolve Color = Color. cases

Inconsistencies are bugs in the compiler, and shoudl be handled in dotnet/roslyn.  Language discussions should be about changing the language (which i would not do here without exceptionally strong reasoning on why it's a good idea).   :)

**Reply by Rekkonnect** [reply]
There is a large overlap between the two, which is why I thought it would be better discussed here initially about properly resolving what the intended behavior is before moving onto Roslyn and fixing the potential bugs and inconsistencies.

Changing the spec is the first part before enabling/disabling specific behaviors that currently stand. If we wish the spec to remain as-is for the time being or forever, at least it may be given as a direction before making adjustments on the compiler to have a clear intention behind the fixes.

**Reply by CyrusNajmabadi** [reply]
> There is a large overlap between the two, which is why I thought it would be better discussed here initially about properly resolving what the intended behavior is

For namespaces, there has never been an intent to support Color/Color afaict.  It's not a thing.  It's explicitly not in the expected set matched here (as it is not a value), and it has no 'type' with which to do the correspondence part.  

> Changing the spec is the first part

No, it def is not.  :)  The impl follows from teh spec, not the other way around.  We only change teh spec *sparingly* in deviation cases where it's felt to be an extremely good idea (which is def lackign here).  

> at least it may be given as a direction before making adjustments on the compiler to have a clear intention behind the fixes.

Right.  That's what i'm trying to do here.  The spec is very clear on this feature and how it is supposed to work.  If the impl is deviating, we have to decide waht is going on *there* (either fixing it, or accepting that it's out of spec and documenting that fact).

**Comment by CyrusNajmabadi** [comment]
> There is a large overlap between the two,

So, as i see it.  There are two impl issues:

1. That we screw this up for constants.  We should fix that in teh impl if possible, or doc it as a deviation if not.
2. That something is going wonky with namespaces (though can you give an example on that).  That part very much surprises me in the first place as i dont' see how Color/Color could ever apply there as we have to do a type-equivalence check... and namespace don't have types.

**Reply by Rekkonnect** [reply]
1. The dotnet/roslyn#68283 PR fixes it, through some apparently unorthodox way, as commented by Aleksey; it will be further looked at if need be.
2. There are some examples with namespaces in the SharpLab repro linked in the thread, clearly showing that it is currently possible to assign a default parameter value with that Color Color pattern. My very inexperienced guess is that the binder successfully ignores the declaring parameter for accessing, resulting in an outer scope binder to resolve an access to the namespace, and perhaps especially special-casing it.

**Reply by CyrusNajmabadi** [reply]
I don't see any sharplab repros in https://github.com/dotnet/roslyn/issues/45739 or https://github.com/dotnet/roslyn/pull/68283.  Can you link to them directly?  Thanks!

**Reply by Rekkonnect** [reply]
In **this** thread I mean: [link](https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoBAGABAYQFcBnAFwHsx8ALCAJ1wF5cEUA6G+gbkxwJIpgAchDABTYgAcIAYzE8MvAMy4xMQmALkANuQYBvXACUxAEwA0uAOJ0xa3AF9MylkgKZ9mXN5YqZ5GDItXQZ8HT0mYL02E1MFH19cf0DSFnYASRhSJTdWSNY2TOykNgBZCAAPADUIbUJ5Lx8EPwCggCFyHTEIGFwOrp7I/u1umDYAFTp6gGVSOlgAc3imlpT+MkpOUIFN2gZmIg2qPbLKmrqGjATG72a0gDZcWwhTAO0AT1xYVMPBEXEpLIxJFfpR/hJpHI2NNKGJ8NoIMRiGV3qCwOd6lxcJgnIosCpkLgACIeG6JViPZ6vGAfNLQuaLNL5dizeYwBZsACiYEkpHey1uBJQlO61NpaPBgKhMPE8MRxHWf1EEKBIJ2wmVUrE0NhcqRhRSPTkAvJwqeorenxlcIRSNw1r1CuYDttyMyZCNl2uVxW5pelqioXC+0DMTMClxmBgmshwIlMaBpJ9gtc9t1rrJnmTCVNIv9NKt6fluHdpE9kRgYgA7gAKACUJu9Od9ySC31wpVR6oxwOYSgALI2fGSEncXcWhGIKqlGAA+XDqbTaIfeEe+hD9jtiUjUcimGvt8d251F/WdtE9utr7xZ5veXEJXGR/GqdSaACCSYSbSGbAA8pW5g4s4L5qBofRfj4AG9rgKBARguJAA==)

> First of all, a simple [repro](https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoBAGABAYQFcBnAFwHsx8ALCAJ1wF5cEUA6G+gbkxwJIpgAchDABTYgAcIAYzE8MvAMy4xMQmALkANuQYBvXACUxAEwA0uAOJ0xa3AF9MylkgKZ9mXN5YqZ5GDItXQZ8HT0mYL02E1MFH19cf0DSFnYASRhSJTdWSNY2TOykNgBZCAAPADUIbUJ5Lx8EPwCggCFyHTEIGFwOrp7I/u1umDYAFTp6gGVSOlgAc3imlpT+MkpOUIFN2gZmIg2qPbLKmrqGjATG72a0gDZcWwhTAO0AT1xYVMPBEXEpLIxJFfpR/hJpHI2NNKGJ8NoIMRiGV3qCwOd6lxcJgnIosCpkLgACIeG6JViPZ6vGAfNLQuaLNL5dizeYwBZsACiYEkpHey1uBJQlO61NpaPBgKhMPE8MRxHWf1EEKBIJ2wmVUrE0NhcqRhRSPTkAvJwqeorenxlcIRSNw1r1CuYDttyMyZCNl2uVxW5pelqioXC+0DMTMClxmBgmshwIlMaBpJ9gtc9t1rrJnmTCVNIv9NKt6fluHdpE9kRgYgA7gAKACUJu9Od9ySC31wpVR6oxwOYSgALI2fGSEncXcWhGIKqlGAA+XDqbTaIfeEe+hD9jtiUjUcimGvt8d251F/WdtE9utr7xZ5veXEJXGR/GqdSaACCSYSbSGbAA8pW5g4s4L5qBofRfj4AG9rgKBARguJAA==).

**Reply by CyrusNajmabadi** [reply]
What is the repro for the namespae issue you are referring to?  That link seems to have a ton of different cases in it.

**Reply by Rekkonnect** [reply]
My bad, I forgot to change the actual repro collection, [this](https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoBAGABAYQFcBnAFwHsx8ALCAJ1wF5cEUA6G+gbkxwJIpgAchDABTYgAcIAYzE8MvAMy4xMQmALkANuQYBvXACUxAEwA0uAOJ0xa3AF9MylkgKZ9mXN5YqZ5GDItXQZ8HT0mYL02E1MFH19cf0DSFnYASRhSJTdWSNY2TOykNgBZCAAPADUIbUJ5Lx8EPwCggCFyHTEIGFwOrp7I/u1umDYAFTp6gGVSOlgAc3imlpT+MkpOUIFN2gZmIg2qPbLKmrqGjATG72a0gDZcWwhTAO0AT1xYVMPBEXEpLIxJFfpR/hJpHI2NNKGJ8NoIMRiGV3qCwOd6lxcJgnIosCpkLgACIeG6JViPZ6vGAfNLQuaLNL5dizeYwBZsACiYEkpHey1uBJQlO61NpaPBgKhMPE8MRxHWf1EEKBIJ2wmVUrE0NhcqRhRSPTkAvJwqeorenxlcIRSNw1r1CuYDttyMyZCNl2uVxW5pelqioXC+0DMTMClxmBgmshwIlMaBpJ9gtc9t1rrJnmTCVNIv9NKt6fluHdpE9kRgYgA7gAKACUJu9Od9ySC31wpVR6oxwOYSgALI2fGSEncXcWhGIKqlGAA+XDqbTaIfeEe+hD9jtiUjUcimFA19vju3Oov6ztont1tfeLPN7y4+/kzelbe70xIQ9ZRVghNyNVHJKsY6rKroopetT1Ne2YJHeT4+K2qQFEUORMswyFZDkpzVJBXr3o+953BuW47nuShfj86pAaqBxUX+2rHsiF7drh0FPnB8F0ihuQoMyBrFNhPYro4ZIEau2Z3GoGi4AA8juYh0JmN4EOYok4s4+KqOomgAIJJgkbRDGwMmVqp2bxgCsYAUqllAsZ8l0BwZmRppUmaG0+k+CZva4CgzmYEAA=) one contains the actual repro with the namespace binding in Method2 and Method3.

**Reply by CyrusNajmabadi** [reply]
Whittling it down, it seems like you're talking about:

```c#
namespace CustomNamespace
{
    public class SomeClass
    {
        public const int MyCustomValue = 34;

        public void Method(int CustomNamespace = CustomNamespace.SomeClass.MyCustomValue)
        {
        }
    }
}
```

?

If so, this is not a `Color Color` case.  This is just due to the fact that parameters are not in scope within their initializers.  So `CustomNamespace.SomeClass.MyCustomValue` binds (correctly) to the namespace, then type, then member.  This appears to be working exactly as expected as per the spec and impl.  I don't see a bug here. 

If that's the case, then the only impl issue identified is hte original.  Namely that constants (either constant-fields, or enum-members) are out of line with the specified behavior of the language.  Currently, i view that as a bug in the impl.  And, unless insurmountable (which i would be surprised by), would not be something we'd change teh spec around as the spec is appropriate in defining the behavior we'd like here. Thanks! :)

**Reply by Rekkonnect** [reply]
Yes, your simplified example is the case with the parameter that I was talking about. Due to the assignment, I considered it would fall under the Color Color case.

Therefore, the cases that do not currently comply with the spec are the const field initializers and the enum field initializers, if I can correctly infer. After settling this, the PR can move forward and focus directly on those cases.

**Reply by CyrusNajmabadi** [reply]
To be clear the `Color Color` case is *not* "hey... there are two things with this name in scope".  The `Color Color` case is: there is a value we see, and the *name* of the value and the *type* of the value are the *same*.  e.g. we see a value that is a field.  And the field's *name* is both "Color" and the field's *type* is "Color".  In that case, we let "Color" bind to both the value (the instance side) and the type (the static side), allowing you to access both instance and static members here.

In the above case, none of this applies.  First, the parameter doesn't fit the `Color Color` case as its type is `Int32`.  Second, we don't even *see* the parameter in the initializer as it's not in scope at all as per the rules of the language.  So it's not even in play.

**Reply by CyrusNajmabadi** [reply]
> Therefore, the cases that do not currently comply with the spec are the const field initializers and the enum field initializers, if I can correctly infer.

Correct.  :)

**Reply by Rekkonnect** [reply]
I appreciate the explanation, moving the PR forward then

**Comment by CyrusNajmabadi** [comment]
Closing out as the spec/lang are workign as intended here.  There appears to be a compiler deviation in the realm of constants (constant fields, and enum members).  We can drive an understanding/decision on what to do about the impl (if anything) in the dotnet/roslyn repo.
